server/apns/apns_sender.go

```
@@ -43,10 +46,47 @@ func NewSenderUsingPusher(pusher Pusher, appTopic string) (connector.Sender, err
 func (s sender) Send(request connector.Request) (interface{}, error) {
 	deviceToken := request.Subscriber().Route().Get(deviceIDKey)
 	logger.WithField("deviceToken", deviceToken).Debug("Trying to push a message to APNS")
-	return s.client.Push(&apns2.Notification{
-		Priority:    apns2.PriorityHigh,
-		Topic:       s.appTopic,
-		DeviceToken: deviceToken,
-		Payload:     request.Message().Body,
-	})
+	push := func() (interface{}, error) {
+		return s.client.Push(&apns2.Notification{
+			Priority:    apns2.PriorityHigh,
+			Topic:       s.appTopic,
+			DeviceToken: deviceToken,
+			Payload:     request.Message().Body,
+		})
+	}
+	withRetry := &retryable{
+		Backoff: backoff.Backoff{
+			Min:    1 * time.Second,
+			Max:    10 * time.Second,
+			Factor: 2,
+			Jitter: true,
+		},
+		maxTries: 3,
+	}
+	return withRetry.execute(push)
```

we should return a specific error `ErrRetryFailed` and based on that error.
in `apns.go` line 96 
`if errSend == ErrRetryFailed {
	newSender, err := NewSender(a.Config)
		if err != nil {
			logger.Warn("APNS Sender could not be recreated")
			return errSenderNotRecreated
		}
		a.SetSender(newSender)
}`   only then to recreate the connection with apns.